core/src/main/java/org/frankframework/receivers/Receiver.java

```diff
@@ -1512,86 +1528,129 @@ private Message compactMessage(Message message, PipeLineSession session) throws
 		return msgBuilder.build();
 	}
 
-	@SuppressWarnings("synthetic-access")
-	public synchronized ProcessResultCacheItem cacheProcessResult(String messageId, String errorMessage, Instant receivedDate) {
-		final ProcessResultCacheItem prci = processResultCache.computeIfAbsent(messageId, key -> {
-			log.debug("{} caching first result for messageId [{}]", this::getLogPrefix, ()->messageId);
-			ProcessResultCacheItem item = new ProcessResultCacheItem();
-			item.receiveDate = receivedDate;
+	/**
+	 * Call this method exactly once per every try to process a message to initialize the internally cached receive-count for the messageId.
+	 * If the listener implements {@link IKnowsDeliveryCount} then the receiveCount will be set to the {@link IKnowsDeliveryCount#getDeliveryCount(RawMessageWrapper)}.
+	 *
+	 * @param rawMessageWrapper The raw message for which to set the receiveCount.
+	 */
+	protected synchronized @Nonnull ProcessStatusCacheItem updateMessageReceiveCount(@Nonnull RawMessageWrapper<M> rawMessageWrapper) {
+		String messageId = Objects.requireNonNull(rawMessageWrapper.getId(), () -> "Message must have an ID! No ID for raw message [" + rawMessageWrapper + "]");
```

Comment:
We actually always had the requirement that messageId may not be NULL at this point in code but it was not verified, and in some unit-tests it was NULL. Make the requirement explicit.